dspace-api/src/main/java/org/dspace/storage/bitstore/BaseBitStoreService.java

```diff
@@ -0,0 +1,224 @@
+/**
+ * The contents of this file are subject to the license and copyright
+ * detailed in the LICENSE and NOTICE files at the root of the source
+ * tree and available online at
+ *
+ * http://www.dspace.org/license/
+ */
+package org.dspace.storage.bitstore;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.security.DigestInputStream;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.Map;
+
+import com.amazonaws.services.s3.model.ObjectMetadata;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.dspace.core.Utils;
+
+/**
+ * BaseBitStoreService base implementation to store
+ * and organize assets in digits.
+ *
+ */
+public abstract class BaseBitStoreService implements BitStoreService {
+
+    protected static Logger log = LogManager.getLogger(DSBitStoreService.class);
+    // Checksum algorithm
+    protected static final String CSA = "MD5";
+    protected static final String MODIFIED = "modified";
+    protected static final String CHECKSUM_ALGORITHM = "checksum_algorithm";
+    protected static final String CHECKSUM = "checksum";
+    protected static final String SIZE_BYTES = "size_bytes";
+
+    // These settings control the way an identifier is hashed into
+    // directory and file names
+    //
+    // With digitsPerLevel 2 and directoryLevels 3, an identifier
+    // like 12345678901234567890 turns into the relative name
+    // /12/34/56/12345678901234567890.
+    //
+    // You should not change these settings if you have data in the
+    // asset store, as the BitstreamStorageManager will be unable
+    // to find your existing data.
+    protected static final int digitsPerLevel = 2;
+    protected static final int directoryLevels = 3;
+
+    /**
+     * Return the intermediate path derived from the internal_id. This method splits
+     * the id into groups which become subdirectories.
+     *
+     * @param internalId The internal_id
+     * @return The path based on the id without leading or trailing separators
+     */
+    protected String getIntermediatePath(String internalId) {
```

Comment:
While not required, it might be nice to add some Unit tests for this refactored method to ensure it's still operating the same as before. There is a lot of code refactoring in this method, and while it looks right, it's hard to tell at a glance.  It'd be better to have tests which verify that for given `internalId` value(s) we end up with the intermediate path we expect.

Reply:
New tests added in `S3BitStoreServiceTest`.